File: src/asset.rs
use std::path::{Path, PathBuf};

use anyhow::Result;

use crate::{
    constants::Paths,
    minify,
    utils::{digest_filename, filename},
    Mode,
};

#[derive(Debug)]
pub struct PublicFile {
    pub path: PathBuf,
    pub prefix: String,
}

#[derive(Debug)]
pub struct Asset {
    pub filename: String,
    pub content: String,
}

impl Asset {
    pub fn from_path(path: &Path) -> Result<Self> {
        let content = std::fs::read_to_string(path)?;
        let filename = filename(path);

        Ok(Self { filename, content })
    }

    pub fn build_css(out_file: &str, in_file: &str, paths: &Paths, mode: Mode) -> Result<Self> {
        let path = PathBuf::from(out_file);
        let source = paths.styles.join(in_file);
        let content = grass::from_path(source, &grass::Options::default())?;

        Ok(match mode {
            Mode::Prod => Self {
                filename: digest_filename(&path, &content),
                content: minify::css(&content.clone())?,
            },
            Mode::Dev => Self {
                filename: filename(path),
                content,
            },
        })
    }
}

File: src/asset.ts
import * as fs from "node:fs/promises";
import * as path from "node:path";
import * as sass from "sass";
import { minify } from "terser";
import { minifyCSS } from "./minify.js";
import { Path } from "./path.js";
import type { Mode, Site } from "./site.js";
import { ensureDir } from "./utils.js";
import type { WriteFromSite } from "./writeable.js";

export interface StaticAsset {
	path: Path;
	prefix: string;
}

export interface Asset extends WriteFromSite {
	source: Path;
	dest: Path;
	write(site: Site): Promise<void>;
}

abstract class AssetBase implements Asset {
	source: Path;
	dest: Path;

	public constructor(source: Path | string, dest: Path | string) {
		this.source = typeof source === "string" ? new Path(source) : source;
		this.dest = typeof dest === "string" ? new Path(dest) : dest;
	}

	abstract build(mode: Mode): Promise<string>;

	public async write(site: Site) {
		const content = await this.build(site.mode);
		await ensureDir(path.dirname(this.dest.absolute));
		await fs.writeFile(this.dest.absolute, content);
	}
}

export class JavaScriptAsset extends AssetBase {
	async build(mode: Mode): Promise<string> {
		const src = await fs.readFile(this.source.absolute, "utf-8");

		if (mode === "dev") {
			return src;
		}

		await this.dest.digest(src);
		const minified = await minify(src);

		if (!minified.code) {
			throw new Error("Could not minify JS");
		}

		return minified.code;
	}
}

export class CssAsset extends AssetBase {
	async build(mode: Mode): Promise<string> {
		const res = sass.compile(this.source.absolute);

		if (mode === "dev") {
			return res.css;
		}

		await this.dest.digest(res.css);
		return minifyCSS(res.css);
	}
}
